controllers/operatorpolicy_controller.go

```diff
@@ -1426,187 +1470,346 @@ func (r *OperatorPolicyReconciler) handleInstallPlan(
 ) (bool, error) {
 	opLog := ctrl.LoggerFrom(ctx)
-	relatedInstallPlans := make([]policyv1.RelatedObject, 0, len(ownedInstallPlans))
-	ipsRequiringApproval := make([]unstructured.Unstructured, 0)
-	anyInstalling := false
-	currentPlanFailed := false
 	complianceConfig := policy.Spec.ComplianceConfig.UpgradesAvailable
+	latestInstallPlan := operatorv1alpha1.InstallPlan{}
 
-	// Construct the relevant relatedObjects, and collect any that might be considered for approval
-	for i, installPlan := range ownedInstallPlans {
-		phase, ok, err := unstructured.NestedString(installPlan.Object, "status", "phase")
-		if !ok && err == nil {
-			err = errors.New("the phase of the InstallPlan was not found")
-		}
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(latestInstallPlanUnstruct.Object, &latestInstallPlan)
+	if err != nil {
+		opLog.Error(err, "Unable to determine the CSV names of the current InstallPlan",
+			"InstallPlan.Name", latestInstallPlanUnstruct.GetName())
 
-		if err != nil {
-			opLog.Error(err, "Unable to determine the phase of the related InstallPlan",
-				"InstallPlan.Name", installPlan.GetName())
-
-			// The InstallPlan will be added as unknown
-			phase = ""
-		}
-
-		// consider some special phases
-		switch phase {
-		case string(operatorv1alpha1.InstallPlanPhaseRequiresApproval):
-			ipsRequiringApproval = append(ipsRequiringApproval, installPlan)
-		case string(operatorv1alpha1.InstallPlanPhaseInstalling):
-			anyInstalling = true
-		case string(operatorv1alpha1.InstallPlanFailed):
-			// Generally, a failed InstallPlan is not a reason for NonCompliance, because it could be from
-			// an old installation. But if the current InstallPlan is failed, we should alert the user.
-			if sub.Status.InstallPlanRef != nil && sub.Status.InstallPlanRef.Name == installPlan.GetName() {
-				currentPlanFailed = true
+		return updateStatus(policy, installPlansNoApprovals), nil
+	}
+
+	ipCSVs := latestInstallPlan.Spec.ClusterServiceVersionNames
+	phase := latestInstallPlan.Status.Phase
+
+	// consider some special phases
+	switch phase { //nolint: exhaustive
+	case operatorv1alpha1.InstallPlanPhaseRequiresApproval:
+	case operatorv1alpha1.InstallPlanPhaseInstalling:
+		return updateStatus(
+			policy,
+			installPlanInstallingCond,
+			existingInstallPlanObj(&latestInstallPlan, string(phase), complianceConfig),
+		), nil
+	case operatorv1alpha1.InstallPlanFailed:
+		// Generally, a failed InstallPlan is not a reason for NonCompliance, because it could be from
+		// an old installation. But if the current InstallPlan is failed, we should alert the user.
+		if len(ipCSVs) == 1 {
+			return updateStatus(
+				policy, installPlanFailed, existingInstallPlanObj(&latestInstallPlan, string(phase), complianceConfig),
+			), nil
+		}
+
+		// If there is more than one CSV in the InstallPlan, make sure this CSV failed before marking it as
+		// noncompliant.
+		for _, step := range latestInstallPlan.Status.Plan {
```

Comment:
@JustinKuli I read the OLM code and it seems that the only successful final states are created (even if it updated) and present. Do you think it's best to do this filtering or should we just mark the InstallPlan as failed even if this operator was successful?

I don't really know how I can get test coverage on this because `clusterServiceVersionNames` order in the InstallPlan are not deterministic so even if I have an InstallPlan with a failing operator, that one may have been processed first and not move on to the successful operator. I suppose I could modify the InstallPlan status to order the expected successful one but that seems quite hacky.

Reply:
I didn't fully understand this part, so if you think it can work without it, that would be fine? I don't like the idea of modifying the InstallPlan status

Reply:
@JustinKuli to clarify, modifying the InstallPlan status would be for test coverage.

So the question is, should an InstallPlan be marked as failed in the OperatorPolicy if the overall InstallPlan failed but the part for the operator being managed by OperatorPolicy succeeded?